library/std/src/sys/unix/process/process_common.rs

```diff
@@ -455,7 +456,31 @@ impl ChildStdio {
 }
 
 impl fmt::Debug for Command {
+    // show all attributes but `self.closures` which does not implement `Debug`
+    // and `self.argv` which is not useful for debugging
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut debug_command = f.debug_struct("Command");
+        debug_command
+            .field("program", &self.program)
+            .field("args", &self.args)
+            .field("env", &self.env)
+            .field("cwd", &self.cwd)
+            .field("uid", &self.uid)
+            .field("gid", &self.gid)
+            .field("saw_nul", &self.saw_nul)
+            .field("groups", &self.groups)
+            .field("stdin", &self.stdin)
+            .field("stdout", &self.stdout)
+            .field("stderr", &self.stderr)
+            .field("pgroup", &self.pgroup);
+
+        #[cfg(target_os = "linux")]
+        {
+            debug_command.field("create_pidfd", &self.create_pidfd);
+        }
+
+        debug_command.finish()?;
+        write!(f, "\nCommand line: ")?;
```

Comment:
This could be extended to also list environment variables and IO redirection.
Many unix shells support doing things like

`FOO=1 BAR=baz /path/to/command arg1 arg1 >out <in 2>err`

If a cwd is set it could be extended further to

`(cd cwd && FOO=1 BAR=baz /path/to/command arg1 arg1 >out <in 2>err)`

On windows it could print the powershell equivalent.

Reply:
Tbh I don't like the idea of keeping a shell-like version of the Command. This is not the purpose of a `Debug` impl and I've only re-added due to the explicit request of bjorn3. It needs special care because users would be tempted to run the command by hand, so we would need to properly escape args as you've commented below, and handle platform specific behaviour. I think it would be best provided by a specific function or by using the `alternate` form.

I would drop this version only to retain the struct-like output which is more conventional. As jyn514 a `Debug` impl spanning multiple lines is also unusual.

Reply:
Not sure if a specific function would any value. In the end it's just a best-effort debug feature. You can't replicate process spawning behavior on a shell with 100% accuracy anyway, e.g. due to file descriptor inheritance. And we probably wouldn't want to commit to supporting it for all platforms since some environments might only have some non-standard shells.

Having both the struct-form and the shell form in one output helps seeing what the shell form is missing, e.g. when a uid is set or something like that.

I do agree with bjorn3 that it is useful to be able to replicate debug output by hand. That's what I often do too when debugging a step in build process that spawns thousands of different commands.

Using the alternate form is reasonable. But I think it should be used for the new format so people don't get grumpy over having their spacebar heating removed.

Reply:
Put the struct output behind the alternate form